include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -352,16 +352,25 @@ __parallel_for(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&&
 
     using __small_submitter = __parallel_for_small_submitter<_ForKernelSmall>;
     using __large_submitter = __parallel_for_large_submitter<_ForKernelLarge, _Ranges...>;
-    // Compile two kernels: one for small-to-medium inputs and a second for large. This avoids runtime checks within a single
-    // kernel that worsen performance for small cases.
-    if (__count < __large_submitter::__estimate_best_start_size(__exec))
+    // Compile two kernels: one for small-to-medium inputs and a second for large. This avoids runtime checks within a
+    // single kernel that worsen performance for small cases. If the number of iterations of the large submitter is 1,
+    // then only compile the basic kernel as the two versions are effectively the same.
+    if constexpr (__large_submitter::__iters_per_work_item > 1)
```

Comment:
I think if you rearrange the large submitter to be the `if` rather than the `else`, you can enable the if statement only when the constexpr is true, and then you can consolidate the two copies of the small submitter invocation into a single one (just using lack of return in place of `else`).

Reply:
Good point, done.